src/lib/tasks/migration_preparation_task.cpp

```diff
@@ -0,0 +1,229 @@
+    }
+  }
+  return true;
+}
+
+int get_node_id(const PolymorphicAllocator<size_t>& alloc) {
+  const auto memsource = dynamic_cast<NUMAMemoryResource*>(alloc.resource());
+  if (memsource) {
+    return memsource->get_node_id();
+  }
+  return -1;
+}
+
+std::vector<ChunkInfo> find_hot_chunks(const StorageManager& storage_manager, const std::chrono::milliseconds& lookback,
+                                       const std::chrono::milliseconds& counter_history_interval) {
+  std::vector<ChunkInfo> chunk_infos;
+  double sum_hottness = 0.0;
+  size_t lookback_samples = lookback.count() / counter_history_interval.count();
+  for (const auto& table_name : storage_manager.table_names()) {
+    const auto& table = *storage_manager.get_table(table_name);
+    const auto chunk_count = table.chunk_count();
+    for (ChunkID i = ChunkID(0); i < chunk_count; i++) {
+      const auto& chunk = table.get_chunk(i);
+      if (chunk_is_completed(chunk, table.chunk_size()) && chunk.has_access_counter()) {
+        const double hottness = static_cast<double>(chunk.access_counter()->history_sample(lookback_samples));
+        sum_hottness += hottness;
+        chunk_infos.emplace_back(ChunkInfo{.table_name = table_name,
+                                           .id = i,
+                                           .node = get_node_id(chunk.get_allocator()),
+                                           .byte_size = chunk.byte_size(),
+                                           .hottness = hottness});
+      }
+    }
+  }
+  std::sort(chunk_infos.begin(), chunk_infos.end(), [](const ChunkInfo& a, const ChunkInfo& b) { return b < a; });
+  return chunk_infos;
+}
+
+std::vector<size_t> count_chunks_by_node(const std::vector<ChunkInfo>& chunk_infos, size_t node_count) {
+  std::vector<size_t> result(node_count);
+  for (const auto chunk_info : chunk_infos) {
+    result.at(chunk_info.node)++;
+  }
+  return result;
+}
+
+MigrationPreparationTask::MigrationPreparationTask(const NUMAPlacementManagerOptions& options) : _options(options) {}
+
+void MigrationPreparationTask::_on_execute() {
+  const auto topology = std::dynamic_pointer_cast<NodeQueueScheduler>(CurrentScheduler::get())->topology();
+
+  auto hot_chunks =
+      find_hot_chunks(StorageManager::get(), _options.counter_history_range, _options.counter_history_interval);
+  size_t chunk_counter = 0;
+  HotNodeSet hot_nodes = find_hot_nodes(get_node_hottness(hot_chunks, topology->nodes().size()));
+
+  std::cout << "Imbalance: " << hot_nodes.imbalance << std::endl;
+  print_vector(hot_nodes.node_hottness, "Hotnesses: ");
+  print_vector(count_chunks_by_node(hot_chunks, topology->nodes().size()), "Chunk counts: ");
+
+  if (hot_nodes.imbalance > _options.imbalance_threshold && hot_nodes.cold_nodes.size() > 0) {
+    std::vector<ChunkInfo> migration_candidates;
+    for (const auto& hot_chunk : hot_chunks) {
+      if (hot_chunk.node < 0 || contains(hot_nodes.hot_nodes, static_cast<NodeID>(hot_chunk.node))) {
```

Comment:
Why could `hot_chunk.node < 0`?

Reply:
I use `-1` to identify nodes that are not pinned to a specific node. I could turn that into a named constant.

Reply:
That would be easier to understand, yes